### DISCUSSION TITLE
[AsyncIO] Best Practices for performing queries in orm event handlers

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#5949] [Usage Questions]
Hello fantastic community,

I am having some issues with the ORM events, particularly `after_update` and `after_insert`. I originally thought that these ran after the changes were committed to the database but I am now learning that they happen when they are added to the session.

Here is my problem:

- An update happens on the User object
- The new User object is added to the session with session.add()
- The after_update event is triggered
-     --> In my event handler, I make another SQL Query
-     --> In the original Async Tornado Handler, I session.commit()
- I get the following exception: `<class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress` from the session.commit() command since the asyncpg is already performing the query in my event handler.

I use those event handlers to fire off events to my frontend subscribers to make them aware of the change that just happened.

Any tips on how I should be handling this?

Thanks!

**Comment by zzzeek** [comment]
> Hello fantastic community,
> 
> I am having some issues with the ORM events, particularly `after_update` and `after_insert`. I originally thought that these ran after the changes were committed to the database but I am now learning that they happen when they are added to the session.

that's not accurate, these events are triggered when a flush() proceeds.  the flush() will occur both as part of the commit(), as well as if called explicitly, or as a result of "autoflush", which is a configurable behavior that causes flush() to invoke before any SQL SELECT statement is to be invoked within the ORM.   if you are seeing after_insert/after_update occur at times you may not expect, look to [autoflush](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#session-flushing) for clarity on why this might be happening. 


> 
> Here is my problem:
> 
>     * An update happens on the User object
> 
>     * The new User object is added to the session with session.add()
> 
>     * The after_update event is triggered
> 
>     * ```
>       --> In my event handler, I make another SQL Query
>       ```
> 
>     * ```
>       --> In the original Async Tornado Handler, I session.commit()
>       ```
> 
>     * I get the following exception: `<class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress` from the session.commit() command since the asyncpg is already performing the query in my event handler.

This sounds like some kind of unexpected concurrency is taking place and would need more specifics to understand why that would be happening.  stack traces are a good place to start.
